### PR TITLE
Add Audio Device Selection to GUI (Fix #376)

### LINKED ISSUE #376: Suggestion for fixing the Audio on Raspbian!
**Environment**

 - OS/Distribution: Raspbian Buster 32bit
 - Desktop Environment:   X11, Pixel
 - Hardware: Tested on RPi4 but I think it might apply to all of them.
 - Exact Version or Commit Hash and Installation Method:  Pull of my current Master (recent copy of yours)

**Describe the bug**
Qt is not able to find the correct Audio output device on its own.
Also Qt is not able to pick the right Audio plugin when both Alsa and PulseAudio are present.
This results in no sound on Raspbian.
My previous Qt5.12 fix was a red herring, the reason it worked was that it didn't have the Pulse Qt plugin, only the Alsa plugin, so qt had to default to Alsa.

**To Reproduce**
1. Build Chiaki as usual on Raspbian *without* Qt5.12
2. Start a game session.
3. No sound will be heard.

**Expected behavior**
Alsa should be available without special Qt hacks. 
Audio should work by default.

**Additional context**
@tomblind discovered that if he removed the /usr/lib/arm-linux-gnueabihf/qt5/plugins/audio/libqtmedia_pulse.so file then Qt defaulted to Alsa and he could hear the sound even with the regular Qt5.

Based on that I went back to the code and now have it working with the regular Qt5 install. No 5.12 version needed.

In streamsession.cpp:   (just to show what does work)

```
       //Commented out as it returns null or nothing on Raspbian.
       //QAudioDeviceInfo audio_device_info(QAudioDeviceInfo::defaultOutputDevice());

	// try pick Alsa if available ("default" on my Raspbian)
	QAudioDeviceInfo audio_device_info;
	QString user_device_name = "default";
	for (QAudioDeviceInfo di : QAudioDeviceInfo::availableDevices(QAudio::AudioOutput))
	{
		//printf("DEVICE:  %s\n", di.deviceName().toLocal8Bit().constData());
		if(di.deviceName() == user_device_name)
		{
			//printf("Will Use Audio Device:  %s\n", di.deviceName().toLocal8Bit().constData());
			audio_device_info = di;
		}
	}
```	
then further down we need to add this device info,    

`audio_output = new QAudioOutput(audio_device_info, audio_format, this);`

I suggest that we add a popup menu in the settings gui where we cross match a premade list of audiodevices maybe even just "default" and "pulse", with what comes up when doing availableDevices(). The intersection would populate the popup menu.
    
"default" should be chosen as default if present.  

That would enable Alsa sound straight away without user having to do anything and people who want to set up with PulseAudio can still do that.

And I could remove Qt5.12 from the install process for Raspian!

Cheers
Fred

### PR DESCRIPTION
**Pull request by Fredrum**
Hi @thestr4ng3r  here's the bits that I just did to add better default handling of Raspian/Linux(?) audio out devices.
There's a selection menu in the Settings.  I don't know where you'd like to have that menu so just move it or tell me where I should move it to?

Not sure if my Git procedure was better this time? :)
Hmm...probably not I only changed 5 files. 
settings.h .cpp
settingsdialog.h .cpp
streamsession.cpp

I'm happy to re-do it if this created a bunch of extra work for you?

Cheers, Fred

### CODE REVIEW

**Comment by thestr4ng3r** [comment]
The f91f2a9 commit looks clean, it should work if you just reset --hard to the current chiaki master, then cherr-pick f91f2a9.

File: gui/src/settings.cpp [hunk 1]
```diff
@@ -106,6 +130,25 @@ unsigned int Settings::GetAudioBufferSize() const
 	return v ? v : GetAudioBufferSizeDefault();
 }
 
+QAudioDeviceInfo Settings::GetAudioOutDevice() const
+{
+	QAudioDeviceInfo outDev;
+
+	QString s = settings.value("settings/audio_out_device", "default").toString();
+	for (QAudioDeviceInfo di : QAudioDeviceInfo::availableDevices(QAudio::AudioOutput))
```
Comment (thestr4ng3r) [inline gui/src/settings.cpp:138]:
```suggestion
	for(QAudioDeviceInfo di : QAudioDeviceInfo::availableDevices(QAudio::AudioOutput))
```

File: gui/src/settings.cpp [hunk 2]
```diff
@@ -106,6 +130,25 @@ unsigned int Settings::GetAudioBufferSize() const
...
+	{
+		if(di.deviceName() == s)
+			outDev = di;
```
Comment (thestr4ng3r) [inline gui/src/settings.cpp:141]: Remove the outDev variable and early-return here:
```suggestion
			return di;
```

File: gui/src/settings.cpp [hunk 3]
```diff
@@ -106,6 +130,25 @@ unsigned int Settings::GetAudioBufferSize() const
...
+	}
+
+	return outDev;
```
Comment (thestr4ng3r) [inline gui/src/settings.cpp:144]: Then return this if nothing else found:
```suggestion
	return QAudioDeviceInfo::defaultOutputDevice();
```

File: gui/src/settingsdialog.cpp [hunk 4]
```diff
@@ -149,6 +151,21 @@ SettingsDialog::SettingsDialog(Settings *settings, QWidget *parent) : QDialog(pa
 	audio_buffer_size_edit->setPlaceholderText(tr("Default (%1)").arg(settings->GetAudioBufferSizeDefault()));
 	connect(audio_buffer_size_edit, &QLineEdit::textEdited, this, &SettingsDialog::AudioBufferSizeEdited);
 
+	audio_device_combo_box = new QComboBox(this);
+	QList<QString> audio_out_device_list = {};
+	for (QAudioDeviceInfo di : QAudioDeviceInfo::availableDevices(QAudio::AudioOutput))
+	{
+		audio_out_device_list.append(di.deviceName());
```
Comment (thestr4ng3r) [inline gui/src/settingsdialog.cpp:158]: No need, for the temporary list, just use `audio_device_combo_box->addItem(di.deviceName(), di.deviceName())` (note the second arg)
Also remove space after `for` and unnecessary `{`s.

File: gui/src/settingsdialog.cpp [hunk 5]
```diff
@@ -149,6 +151,21 @@ SettingsDialog::SettingsDialog(Settings *settings, QWidget *parent) : QDialog(pa
...
+	}	
+	audio_device_combo_box->addItems(audio_out_device_list);
+	auto current_audio_out_device = settings->GetAudioOutDevice();
+	int retrieved_device_index = 0;
+	if(audio_out_device_list.indexOf(current_audio_out_device.deviceName()) >= 0)
+		retrieved_device_index = audio_out_device_list.indexOf(current_audio_out_device.deviceName());
```
Comment (thestr4ng3r) [inline gui/src/settingsdialog.cpp:164]: Use `audio_device_combo_box->findData(current_audio_out_device.deviceName())`.

File: gui/src/settingsdialog.cpp [hunk 6]
```diff
@@ -149,6 +151,21 @@ SettingsDialog::SettingsDialog(Settings *settings, QWidget *parent) : QDialog(pa
...
+	audio_device_combo_box->setCurrentIndex(retrieved_device_index);
+	connect(audio_device_combo_box, SIGNAL(currentIndexChanged(int)), this, SLOT(AudioOutputSelected()));
```
Comment (thestr4ng3r) [inline gui/src/settingsdialog.cpp:166]: Use modern connect syntax, remove `AudioOutputSelected()` and use lambda instead
```suggestion
	connect(audio_device_combo_box, &QComboBox::currentIndexChanged, this, [this, audio_device_combo_box]{
		settings->SetAudioOutDevice(audio_device_combo_box->currentData());
	});
```

File: gui/src/settingsdialog.cpp [hunk 7]
```diff
@@ -149,6 +151,21 @@ SettingsDialog::SettingsDialog(Settings *settings, QWidget *parent) : QDialog(pa
...
+	general_layout->addRow(tr("Audio Out Device"), audio_device_combo_box);
```
Comment (thestr4ng3r) [inline gui/src/settingsdialog.cpp:167]:
```suggestion
	general_layout->addRow(tr("Audio Output Device:"), audio_device_combo_box);
```

File: gui/src/streamsession.cpp [hunk 8]
```diff
@@ -257,8 +295,12 @@ void StreamSession::InitAudio(unsigned int channels, unsigned int rate)
 	audio_format.setSampleSize(16);
 	audio_format.setCodec("audio/pcm");
 	audio_format.setSampleType(QAudioFormat::SignedInt);
+	
+	Settings user_settings;
+	QAudioDeviceInfo audio_device_info = user_settings.GetAudioOutDevice();
```
Comment (thestr4ng3r) [inline gui/src/streamsession.cpp:300]: No need for local user_settings:
```suggestion
	QAudioDeviceInfo audio_device_info = settings->GetAudioOutDevice();
```

**Review by thestr4ng3r — CHANGES_REQUESTED** [review]

File: gui/src/settingsdialog.cpp (hunk 4 above)
Comment (Fredrum) [reply gui/src/settingsdialog.cpp:158]: done

File: gui/src/settingsdialog.cpp (hunk 5 above)
Comment (Fredrum) [reply gui/src/settingsdialog.cpp:164]: done

**Comment by Fredrum** [comment]
Trying to do the reset hard as you suggested.
I'm doing this,

```
git remote set-url origin git@github.com:thestr4ng3r/chiaki.git
git reset --hard origin/master
```

I get these messages, not sure why my other branch 'rpi01' gets mentioned? (git branch -vv  reports 'master')

```
warning: refname 'origin/master' is ambiguous.
warning: refname 'origin/master' is ambiguous.
HEAD is now at 92cd195 Merge branch 'master' of github.com:Fredrum/chiaki into rpi01
```

I tried to just 'accept' your edits, but I did the above steps after that and I don't see the accepted edits in my source files so maybe that got reverted?

Luckily I can see the accepted changes so I can just redo them by hand,

File: gui/src/settingsdialog.cpp (hunk 6 above)
Comment (Fredrum) [reply gui/src/settingsdialog.cpp]: hmm...this change is giving error

```
/home/pi/chiaki/chiaki/gui/src/settingsdialog.cpp:160:80: error: capture of non-variable ‘SettingsDialog::audio_device_combo_box’
  connect(audio_device_combo_box, &QComboBox::currentIndexChanged, this, [this, audio_device_combo_box]{
```

File: gui/src/streamsession.cpp (hunk 8 above)
Comment (Fredrum) [reply gui/src/streamsession.cpp]: I actually tried that earlier but was surprised to get:

```
/home/pi/chiaki/chiaki/gui/src/streamsession.cpp:300:39: error: ‘settings’ was not declared in this scope
  QAudioDeviceInfo audio_device_info = settings->GetAudioOutDevice();

```
StreamSessionConnectInfo  does have a 'session' member.
But I can't see a StreamSessionConnectInfo type member of StreamSession, even if it gets initialized with one. It seems to be just used locally in init?

File: gui/src/settingsdialog.cpp (hunk 6 above)
Comment (thestr4ng3r) [reply gui/src/settingsdialog.cpp]: Remove `, audio_device_combo_box` from the `[...]` and it should work

Comment (Fredrum) [reply gui/src/settingsdialog.cpp]:
```
	connect(audio_device_combo_box, &QComboBox::currentIndexChanged, this, [this]{
		settings->SetAudioOutDevice(audio_device_combo_box->currentData());
	});
```
does not work.  Neither does [=]

Comment (Fredrum) [reply gui/src/settingsdialog.cpp]: does it need a () with something in it straight after the []   ?

Comment (Fredrum) [reply gui/src/settingsdialog.cpp]:
```
connect(audio_device_combo_box, &QComboBox::currentIndexChanged, this, [=](){
		settings->SetAudioOutDevice((QString)audio_device_combo_box->currentData().toString());
	});
```
maybe this got me further but with error:
```
/home/pi/chiaki/chiaki/gui/src/settingsdialog.cpp: In constructor ‘SettingsDialog::SettingsDialog(Settings*, QWidget*)’:
/home/pi/chiaki/chiaki/gui/src/settingsdialog.cpp:162:3: error: no matching function for call to ‘SettingsDialog::connect(QComboBox*&, <unresolved overloaded function type>, SettingsDialog*, SettingsDialog::SettingsDialog(Settings*, QWidget*)::<lambda()>)’
  })
```

Comment (Fredrum) [reply gui/src/settingsdialog.cpp]: This seems a lot harder than it should be. Why is there an overload error when its just a qt native class?
I read other people have issues with that too. (unless I'm doing it wrong)

Comment (thestr4ng3r) [reply gui/src/settingsdialog.cpp]: Yes you are right, coding blind is hard. Fixed now.

**Comment by thestr4ng3r** [comment]
I've cleaned it up and added a few changes on top. Be aware that the device called "default" has nothing to do with `QAudioDeviceInfo::defaultOutputDevice()`, it's just called like that out of one of the plugins.
I also had to make the fetching of available devices async because it took too long and blocked showing the dialog.

**Comment by Fredrum** [comment]
Awesome! Thats great!
Yes i know about "default" its a bit odd but equals Alsa on Raspian. There is no device called Alsa. 
Looking forward to pick this up.

**Pull request merged**